File: Scheduling Algos/FCFS.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.Scanner;

public class FCFS {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter the number of processes: ");
        int n = scanner.nextInt();

        String[] processNames = new String[n]; // Array to store process names
        int[][] processes = new int[n][6]; // Matrix to store process details

        System.out.println("Enter details for each process:");

        // Input for process details in a table format
        System.out.println("ProcessName\tArrival Time\tBurst Time");
        for (int i = 0; i < n; i++) {
            processNames[i] = scanner.next(); // Process Name
            processes[i][0] = scanner.nextInt(); // Arrival Time
            processes[i][1] = scanner.nextInt(); // Burst Time
        }

        // Sort processes based on arrival time while maintaining process IDs order
        Arrays.sort(processes, Comparator.comparingInt(o -> o[0]));

        // Calculate completion, turnaround, waiting, and response times
        int currentTime = 0;
        int totalTurnaroundTime = 0;
        int totalWaitingTime = 0;

        System.out.println("\nGantt Chart:");
        System.out.print(" ");
        for (int i = 0; i < n; i++) {
            System.out.print("---------");
        }
        System.out.println();

        for (int i = 0; i < n; i++) {
            int processID = i;
            currentTime = Math.max(currentTime, processes[processID][0]);

            // Display the Gantt Chart
            for (int j = 0; j < currentTime; j++) {
                System.out.print(" ");
            }
            for (int j = 0; j < processes[processID][1]; j++) {
                System.out.print("|" + processNames[processID]);
            }
            currentTime += processes[processID][1];

            processes[processID][2] = currentTime; // Completion Time
            processes[processID][3] = processes[processID][2] - processes[processID][0]; // Turnaround Time
            processes[processID][4] = processes[processID][3] - processes[processID][1]; // Waiting Time
            processes[processID][5] = processes[processID][4]; // Response Time

            totalTurnaroundTime += processes[processID][3];
            totalWaitingTime += processes[processID][4];
        }

        System.out.println();
        System.out.print(" ");
        for (int i = 0; i < n; i++) {
            System.out.print("---------");
        }
        System.out.println();

        // Display the sorted details according to execution order
        System.out.println("\nExecution Order:\nProcessName\tArrival Time\tBurst Time\tCompletion Time\tTurnaround Time\tWaiting Time\tResponse Time");
        for (int i = 0; i < n; i++) {
            System.out.println(processNames[i] + "\t\t" + processes[i][0] + "\t\t" + processes[i][1] +
                    "\t\t" + processes[i][2] + "\t\t" + processes[i][3] + "\t\t" + processes[i][4] +
                    "\t\t" + processes[i][5]);
        }

        // Calculate and display average turnaround time and average waiting time
        double avgTurnaroundTime = (double) totalTurnaroundTime / n;
        double avgWaitingTime = (double) totalWaitingTime / n;
        System.out.println("\nAverage Turnaround Time: " + avgTurnaroundTime);
        System.out.println("Average Waiting Time: " + avgWaitingTime);

        scanner.close();
    }
}

File: c cpp/Process Scheduling/fcfs.cpp
#include <bits/stdc++.h>
#include "TextTable.h"

using namespace std;

struct Process
{
    int id;
    int AT;
    int BT;
    int CT;
    int TT;
    int WT;
};

bool compareProcessID(const Process &a, const Process &b)
{
    return a.id < b.id;
}

int main()
{
    int numP;
    cout << "Enter number of processes: ";
    cin >> numP;

    vector<Process> processes(numP);

    for (int i = 0; i < numP; i++)
    {
        processes[i].id = i + 1;
        cout << "Enter arrival time for process " << i + 1 << ":";
        cin >> processes[i].AT;
        cout << "Enter burst time for process " << i + 1 << ":";
        cin >> processes[i].BT;
    }

    sort(processes.begin(), processes.end(), [](Process &a, Process &b)
         { return a.AT < b.AT; });

    int currentTime = 0;
    for (int i = 0; i < numP; i++)
    {
        Process &p = processes[i];
        currentTime = max(currentTime, p.AT);
        p.CT = currentTime + p.BT;

        p.TT = p.CT - p.AT;
        p.WT = p.TT - p.BT;

        currentTime = p.CT;
    }

    // Sort processes by process ID
    sort(processes.begin(), processes.end(), compareProcessID);

    TextTable table('-', '|', '+');
    table.add("Process ID");
    table.add("Arrival Time");
    table.add("Burst Time");
    table.add("Completion Time");
    table.add("Turnaround Time");
    table.add("Waiting Time");

    for (int i = 0; i <= numP; i++)
    {
        table.endOfRow();
        table.add(to_string(processes[i].id));
        table.add(to_string(processes[i].AT));
        table.add(to_string(processes[i].BT));
        table.add(to_string(processes[i].CT));
        table.add(to_string(processes[i].TT));
        table.add(to_string(processes[i].WT));
    }
    cout << table;

    // print average turnaround time and average waiting time
    float avg_turnaround_time = 0;
    float avg_waiting_time = 0;

    for (int i = 0; i < numP; i++)
    {
        avg_turnaround_time += processes[i].TT;
        avg_waiting_time += processes[i].WT;
    }

    avg_turnaround_time /= numP;
    avg_waiting_time /= numP;

    cout << "Average Turnaround Time: " << avg_turnaround_time << endl;
    cout << "Average Waiting Time: " << avg_waiting_time << endl;
}